arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/ScopedResource.kt

```diff
@@ -56,23 +60,37 @@ internal class ScopedResource {
     return finalizer?.invoke(ec) ?: Either.Right(Unit)
   }
 
-  suspend fun acquired(finalizer: suspend (ExitCase) -> Unit): Either<Throwable, Boolean> =
-    state.modify { s ->
-      if (s.isFinished()) {
-        // state is closed and there are no leases, finalizer has to be invoked right away
-        Pair(s, suspend {
-          Either.catch {
-            finalizer(ExitCase.Completed)
-            false
-          }
-        })
-      } else {
-        val attemptFinalizer: suspend (ExitCase) -> Either<Throwable, Unit> =
-          { ec -> Either.catch { finalizer(ec) } }
-        // either state is open, or leases are present, either release or `Lease#cancel` will run the finalizer
-        Pair(s.copy(finalizer = attemptFinalizer), suspend { Either.Right(true) })
+  suspend fun acquired(finalizer: suspend (ExitCase) -> Unit): Either<Throwable, Boolean> {
+    val conn = coroutineContext.connection()
+    return state.modify { s ->
+      when {
+        conn.isCancelled() -> {
```

Comment:
We need to check for cancellation while registering the resource into the scope, if while registering the resource a cancellation signal occurs than we need to release the resource immediately with `ExitCase.Cancelled`.

Reply:
Since you're not gonna use the connection anywhere else, wouldn't be better to inline it here? this way you let less space for racing between variable declaration and few functions called in between

Reply:
`isCancelled` can actually be called multiple times here, that's why the connection should be cached.

All operations on `atomic` and `Atomic` that take a function are prone to be called multiple times until it was able to succesfully update using `compareAndSet` since `set(f(get()))` is not an atomic safe operation.

That's also why `Atomic` (and `Ref`) don't take `suspend` (or `F`) signatures. It's advised against to call effectful code inside atomic updates, since slow updates would have a very low chance of a succesful atomic update.
Therefore we also use `modify` here and use a `suspend` lambda to defer the operation to happen outside the atomic update.